modules/dnn/src/ocl4dnn/src/ocl4dnn_pool.cpp

```diff
@@ -0,0 +1,213 @@
+    {
+        kernel_shape_.push_back(i == 0 ? config.kernel.height : config.kernel.width);
+        pad_.push_back(i == 0 ? config.pad.height : config.pad.width);
+        stride_.push_back(i == 0 ? config.stride.height : config.stride.width);
+        im_in_shape_.push_back(config.in_shape[dims - spatial_dims + i]);
+        im_out_shape_.push_back(config.out_shape[dims - spatial_dims + i]);
+    }
+
+    kernel_h_ = kernel_shape_[0];
+    kernel_w_ = kernel_shape_[1];
+    stride_h_ = stride_[0];
+    stride_w_ = stride_[1];
+    pad_h_ = pad_[0];
+    pad_w_ = pad_[1];
+    height_ = im_in_shape_[0];
+    width_ = im_in_shape_[1];
+    pooled_height_ = im_out_shape_[0];
+    pooled_width_ = im_out_shape_[1];
+
+    count_ = 1;
+    for (int i = 0; i < config.out_shape.size(); ++i)
+    {
+        count_ *= config.out_shape[i];
+    }
+}
+
+template<typename Dtype>
+OCL4DNNPool<Dtype>::~OCL4DNNPool()
+{
+    mask_idx_.release();
+}
+
+template<typename Dtype>
+bool OCL4DNNPool<Dtype>::Forward(const UMat& bottom,
+                                 UMat& top,
+                                 UMat& top_mask)
+{
+    bool ret = true;
+    ocl::Queue queue = ocl::Queue::getDefault();
+    size_t global[] = { 128 * 128 };
+    size_t local[] = { 128 };
+    cl_uint argIdx = 0;
+
+    // support 2D case
+    switch (pool_method_)
+    {
+    case LIBDNN_POOLING_METHOD_MAX:
+        {
+            if (top_mask.empty() && mask_idx_.empty())
+            {
+                mask_idx_.create(1, count_, CV_32FC1);
+            }
+            ocl::Kernel oclk_max_pool_forward(CL_KERNEL_SELECT("max_pool_forward"),
+                                              cv::ocl::dnn::ocl4dnn_pooling_oclsrc);
+
+            if (oclk_max_pool_forward.empty())
+                return false;
+
+            argIdx = 0;
+            oclk_max_pool_forward.set(argIdx++, count_);
+            oclk_max_pool_forward.set(argIdx++, ocl::KernelArg::PtrReadOnly(bottom));
+            oclk_max_pool_forward.set(argIdx++, batch_size_);
+            oclk_max_pool_forward.set(argIdx++, channels_);
+            oclk_max_pool_forward.set(argIdx++, height_);
+            oclk_max_pool_forward.set(argIdx++, width_);
+            oclk_max_pool_forward.set(argIdx++, pooled_height_);
+            oclk_max_pool_forward.set(argIdx++, pooled_width_);
+            oclk_max_pool_forward.set(argIdx++, kernel_h_);
+            oclk_max_pool_forward.set(argIdx++, kernel_w_);
+            oclk_max_pool_forward.set(argIdx++, stride_h_);
+            oclk_max_pool_forward.set(argIdx++, stride_w_);
+            oclk_max_pool_forward.set(argIdx++, pad_h_);
+            oclk_max_pool_forward.set(argIdx++, pad_w_);
+            oclk_max_pool_forward.set(argIdx++, ocl::KernelArg::PtrWriteOnly(top));
+            oclk_max_pool_forward.set(argIdx++, mask_idx_.empty() ? 0 : 1);
+            if (mask_idx_.empty())
+                oclk_max_pool_forward.set(argIdx++, NULL);
```

Comment:
`(void*)NULL`, because just NULL has `int` type.

Reply:
fixed in the update